### PR TITLE
keys API

### PR DESCRIPTION
**Pull request by Zil0**
Follows https://github.com/matrix-org/matrix-python-sdk/pull/209 and is based on the work of @pik at https://github.com/pik/matrix-python-sdk/commit/f855cb84ad67713167bdec9825ca6fe03d1d5d24.

This can already be looked at, I tagged it WIP because I didn't add tests yet.

### CODE REVIEW

File: matrix_client/api.py [hunk 1]
```diff
@@ -787,3 +787,119 @@ def set_guest_access(self, room_id, guest_access):
...
+            downloading keys from remote servers.
+        """
+        content = {"one_time_keys": key_request}
+        if timeout:
+            content["timeout"] = timeout
+        return self._send("POST", "/keys/claim", content=content)
+
+    def key_changes(self, from_token, to_token):
+        """Gets a list of users who have updated their device identity keys
+        since a previous sync token.
```
Comment (non-Jedi) [inline matrix_client/api.py]: Should be a single line summary.

**Comment by Zil0** [comment]
Addressed stylistic issues and added tests.

**Comment by non-Jedi** [comment]
@Zil0 could you rebase on master please to make review easier?

**Comment by Zil0** [comment]
done :)

File: matrix_client/api.py [hunk 2]
```diff
@@ -836,3 +836,76 @@ def delete_devices(self, auth_body, devices):
             "devices": devices
         }
         return self._send("POST", "/delete_devices", content=content)
+
+    def upload_keys(self, device_keys={}, one_time_keys={}):
```
Comment (non-Jedi) [inline matrix_client/api.py]: kwargs shouldn't have mutable objects as their default argument. We do this elsewhere in the sdk, and it's wrong. What effectively happens is that the same dictionary (which is mutable) gets reused each time this method is called and could potentially be non-empty in the future.

File: matrix_client/api.py [hunk 3]
```diff
@@ -836,3 +836,76 @@ def delete_devices(self, auth_body, devices):
...
+        """Publishes end-to-end encryption keys for the device.
+
+        Said device must be the one used when logging in.
+
+        Args:
+            device_keys (dict): Optional. Identity keys for the device.
```
Comment (non-Jedi) [inline matrix_client/api.py]: Should give info about the format of this arg in the docstring. Same for below.

File: matrix_client/api.py [hunk 4]
```diff
@@ -836,3 +836,76 @@ def delete_devices(self, auth_body, devices):
...
+            one_time_keys (dict): Optional. One-time public keys.
+        """
+        content = {}
+        if device_keys:
+            content["device_keys"] = device_keys
+        if one_time_keys:
+            content["one_time_keys"] = one_time_keys
+        return self._send("POST", "/keys/upload", content=content)
+
+    def query_user_keys(self, user_id):
```
Comment (non-Jedi) [inline matrix_client/api.py]: This is a higher level method that I don't think belongs in this class.

File: matrix_client/api.py [hunk 5]
```diff
@@ -836,3 +836,76 @@ def delete_devices(self, auth_body, devices):
...
+        Args:
+            user_id (str): The ID of the user whose devices keys to retrieve.
+        """
+        return self.query_keys({user_id: []})
+
+    def query_keys(self, user_devices, timeout=None):
+        """Query HS for public keys by user and optionally device.
+
+        Args:
+            user_devices (dict): The devices whose keys to download. Should be
```
Comment (non-Jedi) [inline matrix_client/api.py]: Please format these for rst as seen in other method docstrings where args need multi-line explanations.

File: matrix_client/api.py [hunk 6]
```diff
@@ -836,3 +836,76 @@ def delete_devices(self, auth_body, devices):
...
+
+    def query_user_keys(self, user_id):
+        """Query HS for public keys of all the given user devices.
+
+        Args:
+            user_id (str): The ID of the user whose devices keys to retrieve.
+        """
+        return self.query_keys({user_id: []})
+
+    def query_keys(self, user_devices, timeout=None):
```
Comment (non-Jedi) [inline matrix_client/api.py]: Where's the `token` parameter for this method? https://matrix.org/speculator/spec/HEAD/client_server/unstable.html#post-matrix-client-r0-keys-query

File: matrix_client/api.py [hunk 7]
```diff
@@ -836,3 +836,76 @@ def delete_devices(self, auth_body, devices):
...
+        content = {"one_time_keys": key_request}
+        if timeout:
+            content["timeout"] = timeout
+        return self._send("POST", "/keys/claim", content=content)
+
+    def key_changes(self, from_token, to_token):
+        """Gets a list of users who have updated their device identity keys.
+
+        Args:
+            from_token (str): The desired start point of the list.
```
Comment (non-Jedi) [inline matrix_client/api.py]: Please explicitly state that this should be taken from `next_batch` field of a sync response.

File: test/api_test.py [hunk 8]
```diff
@@ -148,3 +148,75 @@ def test_delete_devices(self):
...
+            "@alice:example.com": {
+                "ed25519:JLAFKJWSCS": ("dSO80A01XiigH3uBiDVx/EjzaoycHcjq9lfQX0uWsqxl2gi"
+                                       "MIiSPR8a4d291W1ihKJL/a+myXS367WT6NAIcBA")
+            }
+        }
+    }
+
+    @responses.activate
+    @pytest.mark.parametrize("args", [
+        {},
```
Comment (non-Jedi) [inline test/api_test.py]: I don't really entirely understand pytest. Where is this empty dict going? `upload_keys` only takes two arguments, not three.

File: matrix_client/api.py (hunk 4 above)
Comment (Zil0) [reply matrix_client/api.py]: right, same for claim_key then?

File: test/api_test.py (hunk 8 above)
Comment (Zil0) [reply test/api_test.py]: https://docs.pytest.org/en/latest/parametrize.html  
The test is called 3 times, one with {} which stands for no argument, then with device_keys set, then with one_time_keys. Not sure why I didn't do a fourth test with both set. Then we call the API method with the arguments found in `args`.  
If this is too confusing for what it is, I can go back to copy-pasting, which won't make much difference seeing the rest of the file :)

**Comment by Zil0** [comment]
Making the changes but before I forget again:  
Signed-off-by: Valentin Deniaud \<valentin.deniaud@inpt.fr\>

File: matrix_client/api.py (hunk 6 above)
Comment (Zil0) [reply matrix_client/api.py]: Good catch, I wasn't looking at unstable :)

File: matrix_client/api.py (hunk 4 above)
Comment (Zil0) [reply matrix_client/api.py]: assumed it was

**Comment by Zil0** [comment]
fixed :)

**Comment by non-Jedi** [comment]
LGTM. Thanks!

**Pull request merged**